migrate/20241203_authorization_refactor.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+Sequel.migration do
+  up do
+    create_table(:action_type) do
+      uuid :id, primary_key: true
+      String :name, unique: true, null: false
+    end
+    # Action Type UUID primary keys created via:
+    # UBID.from_parts(i, UBID::TYPE_ACTION_TYPE, 0, i).to_uuid
+    from(:action_type).import([:id, :name],
+      [["00000000-0001-835a-8000-000000000001", "Project:delete"],          # 1
+        ["00000000-0002-835a-8000-000000000002", "Project:edit"],
+        ["00000000-0003-835a-8000-000000000003", "Project:view"],
+        ["00000000-0004-835a-8000-000000000004", "Project:user"],
+        ["00000000-0005-835a-8000-000000000005", "Project:billing"],
+        ["00000000-0006-835a-8000-000000000006", "Project:github"],
+        ["00000000-0007-835a-8000-000000000007", "Vm:create"],
```

Comment:
oh man, it's going to bother me for no good reason in particular when we add another project action (or whatever) and it is numerically non-contiguous

Reply:
We don't have to use sequential integer values.  We could have "Project"` actions be 1xxx, "Vm" actions be 2xxx, etc.  Let me know if you prefer that approach.

Reply:
How about encoding them look good in base32/ubid? like `vm0create` (other separator suggestions welcome, the 0, is an homage to null termination, but may also be one of the weird characters to prevent illegible ubids re: 0/O)

Reply:
That's a great idea, it will debugging cases with pry a lot easier.  I'll work on that.

Reply:
Hmm I think I may have erred by thinking of  `0`, because of how "o" folds into it, and "o" is a vowel...that is, relatively common. Do any other characters stand out better as delimiters vs data? Say, a rare one, like "z". I'm also not sure what works best visually.

Reply:
But possibly, it doesn't get better than this. As I'm squinting at it it occurred to me. I also wonder if trying to eliminate entropy by having exactly one copy of the in-band signaling is superior, e.g.

Original: 

`tt0ps0c0nn021g0ps0c0nnect1`

zeroed:

`tt0000000000000ps0c0nnect1`

For a quick preview of `z`
`ttzzzzzzzzzzzzzpszc0nnect1`

I suppose maybe `0` is good in spite of being seen in some data, because it's a tall character. It stands out.

And the couple of bytes that ubids needs to be uuids will not be quite as harmonious as all-zeros.

There's also the possibility of shifting the in-band data to the left to pad it from the parity bits at the end, e.g. more like:

`tt0ps0c0nnect000000000001`

What do you think? I think talked (or rather, read) myself out of z-padding but curious what you think of decreasing repetition and effect on legibility, and left-vs-right alignment.

Reply:
> But possibly, it doesn't get better than this. As I'm squinting at it it occurred to me. I also wonder if trying to eliminate entropy by having exactly one copy of the in-band signaling is superior, e.g.
> 
> Original:
> 
> `tt0ps0c0nn021g0ps0c0nnect1`
> 
> zeroed:
> 
> `tt0000000000000ps0c0nnect1`

This is an invalid ubid:

```ruby
udec("tt0000000000000ps0c0nnect1")
UBIDParseError: Invalid bottom bits parity
```

> For a quick preview of `z` `ttzzzzzzzzzzzzzpszc0nnect1`

This is also an invalid ubid:

```ruby
udec("ttzzzzzzzzzzzzzpszc0nnect1")
UBIDParseError: Invalid top bits parity
```

> I suppose maybe `0` is good in spite of being seen in some data, because it's a tall character. It stands out.
> 
> And the couple of bytes that ubids needs to be uuids will not be quite as harmonious as all-zeros.
> 
> There's also the possibility of shifting the in-band data to the left to pad it from the parity bits at the end, e.g. more like:
> 
> `tt0ps0c0nnect000000000001`
> 
> What do you think? I think talked (or rather, read) myself out of z-padding but curious what you think of decreasing repetition and effect on legibility, and left-vs-right alignment.

Personally, I don't have strong feelings either way.  However, there is less room at the left side than the right side, as the middle of the UBID requires certain things. So if you are going to only have it in a single place, it would be better to have it on the left side.  I did it in both places to make it possible to read from either side, though the 

Here's how it could look on the right hand side only:

```
tt00000000021g0ps0c0nnect1
ttzzzzzzzz021g0ps0c0nnect1
```

Another option would be putting part on the left and part on the right.  This would allow for  longer names for types, instead of the character abbreviation:

```
ttpr1vsvbn020g0zzzc0nnect0
ttpr1vsvbn020g0d1sc0nnect0
```

What would you prefer?

Reply:
to me, the right-hand-side only  tracks better. Don't we have a list of idiomatic abbreviations (the first two bytes) for nearly everything in ubids anyway? This could be the rule for how to abbreviate the object type.

I kinda like your idea of a hybrid, with a bunch of "z"s for padding and `0` for delimiter:

> ttzzzzzzzz021g0ps0c0nnect1

Reply:
Change made: https://github.com/ubicloud/ubicloud/commit/c5ccc981bee4a9714704fcc8c583c778b905e14b#diff-b6e51bff148ac6a959b8da686a14246830c7f6f193704f16f848753cb2b088d0R12

A nice result of this is the UUIDs are also recognizable as different from random UUIDs (e.g. `ffffffff-ff00-835a-802d-903439602b50`).

I noticed the ActionType/ActionTag vanity prefixes do not match the UBID prefixes in some cases:

* Project: `pr` instead of `pj`
* SubjectTag: `st` instead of `ts`
* ActionTag: `at` instead of `ta`
* ObjectTag: `0t` instead of `t0`

If you'd like the prefixes to be made consistent, let me know and I can make that change.

Reply:
I also realized I only used `z` on the left side, the right side is still padded with `0` (or `1` if the first character in the vanity is `0`).  Please let me know if you would like to change that.

Reply:
Upon further review, I can convert two additional bytes from `0` to `z`:

```
tazzzzzzzzz20gzps0d1sc0nn1
tazzzzzzzzz20gzzzzps0v1ew0
```

Should I use that format?

Reply:
Or maybe it would be preferable to always have the type in the same position?:

```
tazzzzzzzzz20qzd1sc0nn0ps0
tazzzzzzzzz20qzzzzv1ew0ps1
```

Reply:
Interesting idea. I think I'm mostly out of good ideas now, it does seem a bit weird to have to read right to left; the type bytes are themselves a bit familiar and help me track the progression.

I do think keeping it as `0` on the right-hand-side works pretty well, in that I think long runs of `0` dilute their value as a (unusually tall) separator, whereas `z` help you zoom past the boring parts (it's not beyond me that `z` is in `zoom` as well as `zzz` is sometimes a reaction to boring things)

Reply:
that's to say, personally, I like `ttzzzzzzzz021g0ps0c0nnect1` (with the leading `0` before `ps` to separate)

Reply:
OK.  Changes made: https://github.com/ubicloud/ubicloud/commit/d3f0c03e40d68b07efdd49ed92038016e06e0f87#diff-b6e51bff148ac6a959b8da686a14246830c7f6f193704f16f848753cb2b088d0R13

So here are the rules as I understand them:

* First 2 characters: UBID type (required), either `tt` or `ta` currently
* Next 12 characters:  `zzzzzzzz021g`
* 0-9 padding characters of `z`
* `0` delimiter if there is a type
* 2 characters for embedded type (UBID type, e.g. `pj` for Project), if there is a type
* `0` delimiter
* 1-7 characters for action/tag name
* Last character: `0` or `1` (depending on parity)

Examples:

* PrivateSubnet:disconnect: `ttzzzzzzzz021g0ps0d1sc0nn0`
* Project:view: `ttzzzzzzzz021gzzz0pj0v1ew0`
* Postgres:Firewall:edit: `ttzzzzzzzz021gzzz0pf0ed1t1`
* Member global tag: `tazzzzzzzz021gzzzz0member0`

Note that `Postgres:Firewall` action types use the `pf` prefix, because they need to be different than `Postgres` actions (which use `pg`), but the UBID `pf` prefix is used for Postgres::FirewallRule, which is slightly different.  Authorizations for `Postgres:Firewall` actions appear to be against `pg.id`, not the id for specific firewall rules.  Maybe it would be better to change:

* `Postgres:Firewall:edit` to `Postgres:fwedit`
* `Postgres:Firewall:view` to `Postgres:fwview`

What do you think?

Reply:
I made the Postgres:Firewall permission name changes.  One of the commits changes the application to use the new names.